File: ureka-framework-java/app/src/main/java/ureka/framework/logic/stage_worker/MsgSender.java
package ureka.framework.logic.stage_worker;

import ureka.framework.Environment;
import ureka.framework.model.SharedData;
import ureka.framework.model.message_model.Message;
import ureka.framework.model.message_model.RTicket;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.communication.bluetooth.BluetoothService;
import ureka.framework.resource.logger.SimpleLogger;
import ureka.framework.resource.logger.SimpleMeasurer;

import java.util.HashMap;
import java.util.Map;

public class MsgSender {
    private SharedData sharedData;
    private MeasureHelper measureHelper;
    public MsgSender(SharedData sharedData, MeasureHelper measureHelper) {
        this.sharedData = sharedData;
        this.measureHelper = measureHelper;
    }

    // Resource (Simulated Comm)
    public void startSimulatedComm() {
        this.sharedData.setSimulatedCommCompletedFlag(false);
    }

    public void completeSimulatedComm() {
        this.sharedData.setSimulatedCommCompletedFlag(true);
    }

    public void waitSimulatedCommCompleted() throws InterruptedException {
        if (Environment.DEPLOYMENT_ENV.equals("TEST")) {
            while (!this.sharedData.getSimulatedCommCompletedFlag()) {
                Thread.sleep((long) Environment.SIMULATED_COMM_INTERRUPT_CYCLE_TIME);
            }
        } else if (Environment.DEPLOYMENT_ENV.equals("PRODUCTION")) {
            this.sharedData.getSimulatedCommReceiverThread().join();
        }
    }

    // Resource (Bluetooth Comm)
    public void startBluetoothComm() {
        this.sharedData.setBluetoothCommCompletedFlag(false);
    }

    public void completeBluetoothComm() {
        this.sharedData.setBluetoothCommCompletedFlag(true);
    }

    // something to fixed
    public void connectBluetoothComm() throws Exception {
        BluetoothService.ConnectingWorker connectingWorker = new BluetoothService.ConnectingWorker(
                BluetoothService.SERVICE_UUID,
                BluetoothService.SERVICE_NAME,
                BluetoothService.RECONNECT_TIMES,
                BluetoothService.RECONNECT_INTERVAL
        );
        this.sharedData.setConnectionSocket(connectingWorker.connect());
    }

    public void closeBluetoothConnection() throws Exception {
        this.sharedData.getConnectionSocket().close();
    }

    // [STAGE: (S)] Send Message
    public void sendXxxMessage(String messageOperation, String messageType, String sentMessageJson) {
        SimpleMeasurer.measureWorkerFunc(this::_sendXxxMessage, messageOperation, messageType, sentMessageJson); // something to fixed
    }
    public void _sendXxxMessage(String messageOperation, String messageType, String sentMessageJson) throws Exception {
        // Generate Message
        if ((messageOperation.equals(Message.MESSAGE_RECV_AND_STORE) ||
                messageOperation.equals(Message.MESSAGE_VERIFY_AND_EXECUTE)) &&
                (messageType.equals(UTicket.MESSAGE_TYPE) ||
                        messageType.equals(RTicket.MESSAGE_TYPE))) {

            Map<String, String> messageRequest = new HashMap<>();
            messageRequest.put("messageOperation", messageOperation);
            messageRequest.put("messageType", messageType);
            messageRequest.put("messageStr", sentMessageJson);

            try {
                Message newMessage = new Message(messageRequest);
                String newMessageJson = Message.messageToJsonstr(newMessage); // something to fixed
                // SimpleLogger.log("debug", "sentMessageJson: " + sentMessageJson);
            } catch (IllegalArgumentException error) { // pragma: no cover -> Weird M-Request
                throw new RuntimeException("Weird M-Request: " + error);
            }
        } else { // pragma: no cover -> Weird M-Request
            throw new RuntimeException("Weird M-Request");
        }

        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            SimpleLogger.simpleLog("info",
                    "+ " + this.sharedData.getThisDevice().getDeviceName() +
                            " is sending message to " +
                            this.sharedData.getSimulatedCommChannel().getEnd().getSharedData().getThisDevice().getDeviceName() + "...");

            // Simulate Network Delay
            for (int i = 0; i < Environment.SIMULATED_COMM_DELAY_COUNT; i++) {
                SimpleLogger.simpleLog("info", "+ network delay");
                SimpleLogger.simpleLog("info", "+ network delay");
                SimpleLogger.simpleLog("info", "+ network delay");
                if (Environment.DEPLOYMENT_ENV.equals("PRODUCTION")) { // pragma: no cover -> PRODUCTION
                    Thread.sleep((long) Environment.SIMULATED_COMM_DELAY_DURATION);
                }
            }

            this.sharedData.getSimulatedCommChannel().getSenderQueue().put(newMessageJson);
        } else { // pragma: no cover -> PRODUCTION
            SimpleLogger.simpleLog("info",
                    "+ " + this.sharedData.getThisDevice().getDeviceName() +
                            " is sending message to BT_address or BT_name...");

            this.sharedData.getConnectionSocket().sendMessage(newMessageJson);
        }
    }
}

File: ureka-framework-main/ureka_framework/logic/stage_worker/msg_sender.py
# Deployment Environment
from ureka_framework.environment import Environment

# Data Model (RAM)
from pydantic import ValidationError
from ureka_framework.model.shared_data import SharedData
import ureka_framework.model.message_model.message as message
from ureka_framework.model.message_model.message import Message, message_to_jsonstr
import ureka_framework.model.message_model.u_ticket as u_ticket
import ureka_framework.model.message_model.u_ticket as r_ticket

# Resource (Bluetooth Comm)
try:
    HAS_PYBLUEZ = True
    import ureka_framework.resource.communication.bluetooth.bluetooth_service as bt_service
    from ureka_framework.resource.communication.bluetooth.bluetooth_service import (
        AcceptSocket,
        ConnectingWorker,
        ConnectionSocket,
    )
except ImportError:
    HAS_PYBLUEZ = False
    # raise RuntimeError(
    #     "PyBlueZ not found - only support SIMULATED comm but not BLUETOOTH comm"
    # )

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Resource (Measurer)
from ureka_framework.resource.logger.simple_measurer import measure_worker_func

# Measure Helper
from ureka_framework.logic.stage_worker.measure_helper import MeasureHelper

# Threading
import time


class MsgSender:
    def __init__(
        self,
        shared_data: SharedData,
        measure_helper: MeasureHelper,
    ) -> None:
        self.shared_data = shared_data
        self.measure_helper = measure_helper

    ######################################################
    # Resource (Simulated Comm)
    ######################################################
    def start_simulated_comm(self) -> None:
        self.shared_data.simulated_comm_completed_flag = False

    def complete_simulated_comm(self) -> None:
        self.shared_data.simulated_comm_completed_flag = True

    def wait_simulated_comm_completed(self) -> None:
        if Environment.DEPLOYMENT_ENV == "TEST":
            # Pytest terminates all daemon threads when main thread is finished
            while not self.shared_data.simulated_comm_completed_flag:
                time.sleep(Environment.SIMULULATED_COMM_INTERRUPT_CYCLE_TIME)
        elif Environment.DEPLOYMENT_ENV == "PRODUCTION":
            self.shared_data.simulated_comm_receiver_thread.join()

    ######################################################
    # Resource (Bluetooth Comm)
    ######################################################
    def start_bluetooth_comm(self) -> None:
        self.shared_data.bluetooth_comm_completed_flag = False

    def complete_bluetooth_comm(self) -> None:
        self.shared_data.bluetooth_comm_completed_flag = True

    def connect_bluetooth_comm(self) -> None:
        self.shared_data.connecting_worker = ConnectingWorker(
            service_uuid=bt_service.SERVICE_UUID,
            service_name=bt_service.SERVICE_NAME,
            reconnect_times=bt_service.RECONNECT_TIMES,
            reconnect_interval=bt_service.RECONNECT_INTERVAL,
        )
        self.shared_data.connection_socket: ConnectionSocket = (
            self.shared_data.connecting_worker.connect()
        )

    def close_bluetooth_connection(self) -> None:
        self.shared_data.connection_socket.close()

    ######################################################
    # [STAGE: (S)] Send Message
    ######################################################
    @measure_worker_func
    def _send_xxx_message(
        self, message_operation: str, message_type: str, sent_message_json: str
    ) -> None:
        # Generate Message
        if (
            message_operation == message.MESSAGE_RECV_AND_STORE
            or message.MESSAGE_VERIFY_AND_EXECUTE
        ) and (message_type == u_ticket.MESSAGE_TYPE or r_ticket.MESSAGE_TYPE):
            message_request: dict = {
                "message_operation": f"{message_operation}",
                "message_type": f"{message_type}",
                "message_str": f"{sent_message_json}",
            }
            try:
                new_message = Message(**message_request)
                new_message_json = message_to_jsonstr(new_message)
                # simple_log("debug", f"sent_message_json: {sent_message_json}")
            except ValidationError as error:  # pragma: no cover -> Weird M-Request
                raise RuntimeError(f"Weird M-Request: {error}")
        else:  # pragma: no cover -> Weird M-Request
            raise RuntimeError("Weird M-Request")

        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            simple_log(
                "info",
                f"+ {self.shared_data.this_device.device_name} is sending message "
                f"to {self.shared_data.simulated_comm_channel.end.shared_data.this_device.device_name}...",
            )

            # Simulate Network Delay
            for _ in range(Environment.SIMULULATED_COMM_DELAY_COUNT):
                simple_log("info", f"+ network delay")
                simple_log("info", f"+ network delay")
                simple_log("info", f"+ network delay")
                if (
                    Environment.DEPLOYMENT_ENV == "PRODUCTION"
                ):  # pragma: no cover -> PRODUCTION
                    time.sleep(Environment.SIMULULATED_COMM_DELAY_DURATION)

            self.shared_data.simulated_comm_channel.sender_queue.put(new_message_json)
        else:  # pragma: no cover -> PRODUCTION
            simple_log(
                "info",
                f"+ {self.shared_data.this_device.device_name} is sending message "
                f"to BT_address or BT_name...",
            )

            self.shared_data.connection_socket.send_message(new_message_json)
